server: Use a strict CSP for all pages

fmeum opened the pull request:
The CSP allows at most the current origin (and subdomains) as well as a few inline scripts pinned by hashes.

<!-- Optional: Provide additional context (beyond the PR title). -->

<!-- Optional: link a GitHub issue.
     Example: "Fixes #123" will auto-close #123 when the PR is merged. -->

**Related issues**: buildbuddy-io/buildbuddy-internal#3911

fmeum commented on server/http/interceptors/interceptors.go:
```diff
@@ -37,15 +38,41 @@ import (
+		scriptSrcs = append(scriptSrcs, fmt.Sprintf("'sha256-%s'", base64.StdEncoding.EncodeToString(scriptHash[:])))
+	}
+	return strings.Join([]string{
+		"default-src 'self'",
```
I clicked around in the app and it worked, but please double-check this list and try to think of other kinds of access it may need.

bduffany commented on server/http/interceptors/interceptors.go:
```diff
@@ -37,15 +38,41 @@ import (
+		"style-src 'self' https://fonts.googleapis.com/css",
+		// The fixed hash is for the inline script that injects the Google Tag Manager. Additional hashes can be
+		// added by the caller.
+		"script-src 'self' https://www.googletagmanager.com 'sha256-3dO3w7Y+Pvr8nNuM9JioBpuBGrxgMBNZhPSqYDD4qn4=' " + strings.Join(scriptSrcs, " "),
```
nit: wdyt about using a [nonce-based](https://web.dev/articles/strict-csp#structure) approach instead of hash-based? I am not an expert in CSP but it seems like nonce-based might be a little easier to maintain since we don't have to update these hashes if the contents change

fmeum replied:
A hash-based approach generally has slightly better security guarantees (think `</script>` injection into server-generated data), but they don't apply to our current usage. It also makes the page itself static, which can help with caching.

A nonce-based approach doesn't require updating the hashes, which are currently only two which are probably very stable, but you are right that this could turn out to be more maintainable in the future. 

It's a close call for me. I did update the PR to use the nonce-based approach since it makes it easier for users to customize the index page: they just need to add `{{.Nonce}}` to their `script` tags rather than having to adding hashes to the Go constant.

fmeum commented on server/static/static.go:
```diff
@@ -213,12 +216,14 @@ func serveIndexTemplate(ctx context.Context, env environment.Env, tpl *template.
 		http.Error(w, err.Error(), http.StatusInternalServerError)
 		return
 	}
+	nonce := interceptors.SetContentSecurityPolicy(w.Header())
```
Yes, this was wastefully generating two nonces for the index page, with only the latter CSP header being sent. I am now storing the nonce in the context, which is indeed much nicer. Thanks for the suggestion!

fmeum commented:
@bduffany I added reporting to a new endpoint and also switched to report only mode when I observed that we may be using a [dep that unnecessarily uses `eval`](https://github.com/protobufjs/protobuf.js/issues/1754). We can enable this in dev and react to reports (or decide to weaken the policy).

bduffany commented on server/http/csp/csp.go:
```diff
@@ -0,0 +1,23 @@
+// The associated value is always of type template.HTMLAttr.
+type Nonce struct{}
+
+const ReportingEndpoint = "/csp-report"
```
it makes me a little nervous that anyone could `curl` this endpoint and trigger an alert for us - maybe just `log.CtxInfo()` for now?

fmeum replied:
Updated to `log.CtxWarning`

bduffany replied:
Warning-level logs are considered errors in our canarying process, and can trigger the apps to be rolled back if too many warnings are logged - we might want to use `Info` until we're more confident that there shouldn't be any CSP violations

fmeum replied:
TIL, thanks!

bduffany reviewed (approved):
LGTM (if we change Warning => Info)
